### ISSUE TITLE
Partially complete uploads can mislead

### ISSUE THREAD

**Issue by RichardScothern** [docker/distribution#936]
It is possible for a `docker push` to not complete entirely leaving the repository in a bad state, for example due to S3 flakiness.  This is because storing the manifest and tagging it are not atomic operation.  Even though an error is returned by the Registry, no cleanup is attempted.

When a `tags/list` is performed on `dockerhubenterprise/trusted-registry-dev`, a tag `asdf` is included in the output.  Attempting to pull with this tag fails because there is no file at
`/docker/registry/v2/repositories/dockerhubenterprise/trusted-registry-dev/_manifests/tags/latest/current/link`.

There are a couple of solutions:
Improve `tags/list` to check for the presence of the link file.

Scan the registry for these inconsistencies and report/fix them (probably covered by #309)

cc: @BrianBland

**Comment by stevvooe** [comment]
We need to be careful about creating a performance bottleneck when checking for tags. Because of the backend storage we are using, we are kind of left with possible inconsistencies in returns from the API. Effectively, `tags/list` and `_catalog` are best effort. I am not sure we can truly fix this until we move to a consistent store for metadata.

What is the main effect of listing a tag that might be there? Are we seeing bugs or failures due to this inconsistency?

**Comment by stevvooe** [comment]
@RichardScothern is this possibly related to https://github.com/docker/distribution/issues/938?

**Comment by RichardScothern** [comment]
This isn't the same as #938 as this is the hub registry, which is v2.0.1 (pre delete) functionality.  It might be the same as #892.

Agreed with the extra IO cost on listing tags.  I think a cleanup operation is the right thing to do here.

**Comment by dmp42** [comment]
@stevvooe yes, we have (at least one) stuff on the hub that appears in the tag list but is un-pullable.

**Comment by stevvooe** [comment]
@RichardScothern @dmp42 Can we cleanup safely without locking the backend?

I'm really starting to think we need to move metadata out of the backend. The constraints are problematic and we are running into more bugs.

**Comment by RichardScothern** [comment]
I'll preface this by saying I totally agree with @stevvooe 's second point above.

I'm changing my mind on this one: soft-deleting a manifest by digest can leave a dangling tag (as pointed out by @dmp42 above), which will still appear in `tags/list`.

Without removing more files in soft delete (and creating a situation where only a subset of files are successfully removed) the only way i can see around this is for 
tag listing code to check if the hash representing each tag had a corresponding link file within the manifest hierarchy.  The cost of this is one  `stat` call per tag.  I believe this will create 2n s3 requests rather than n (where n is the number of tags).

**Comment by RichardScothern** [comment]
I did an experiment with the above implementation (checking for missing link file for each tag when listing tags).

List is actually implemented by 1 request to S3 (my above assumption was incorrect).  The fix requires `tags/list` requires N S3 roundtrips, which on large tag ordinals (some repositories have thousands of tags) is prohibitive.

**Comment by stevvooe** [comment]
@RichardScothern Another way around this is to have drivers implement a `Find` command. Basically, this let's the caller specify a prefix and they get all files under that prefix, avoiding the extra round trips to walk a hierarchy.

**Comment by BrianBland** [comment]
Alternatively, add a `recursive` flag to the driver `List` command

**Comment by RichardScothern** [comment]
However find is implemented, the cost has to be paid at the time of calling it.  By either listing tags (by checking that the digest pointed to by the tag has not been soft deleted) or at manifest delete time (by finding the tag corresponding to the digest and removing it).

I did some experiments (from my laptop to S3 on the wired network):

Currently, listing tags can be served for any tag ordinal in 1 i/o and takes ~700ms

Using `Find` to gather enough information to either delete the tag or check the tag has been deleted:

For `library/ubuntu` (~50 tags) this would take ~1100ms

For a repository with 1800 tags this operation would take 7000ms

The largest tag ordinal on the hub is ~4k.
